### PR TITLE
feat!: versioned namespaces

### PR DESCRIPTION
**Pull request by rootulp**
Part of https://github.com/celestiaorg/celestia-app/issues/1434
Closes https://github.com/celestiaorg/celestia-app/issues/1489

### CODE REVIEW

**Comment by evan-forbes** [comment]
Mind elaborating on how this is apphash breaking to speed up reviews?

**Review by mojtaba-esk — APPROVED** [review]
LGTM.
There are a few linter complains as well.

**Comment by rootulp** [comment]
Good question. AppHash is a hash of a block when serialized into a byte array. I assumed the AppHash would be different because the Block contains Data. Data contains Txs. Txs contain BlobTxs. BlobTxs contain the Cosmos SDK txs with a MsgPayForBlobs with namespace IDs that are larger. So I think this change may only indirectly modify AppHash.

File: go.mod [hunk 1]
```diff
@@ -185,5 +185,6 @@ require (
 replace (
 	github.com/cosmos/cosmos-sdk => github.com/celestiaorg/cosmos-sdk v1.8.0-sdk-v0.46.7
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-	github.com/tendermint/tendermint => github.com/celestiaorg/celestia-core v1.15.0-tm-v0.34.23.0.20230308212354-07aba19d0bda
+	// TODO remove this replace once we merge dependent PR and cut a celestia-core release
+	github.com/tendermint/tendermint => ../celestia-core
```
Comment (evan-forbes) [inline go.mod:189]: can we point to a commit here so we can see failing tests here and in our own local repos?

File: pkg/appconsts/appconsts.go [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
 	"github.com/tendermint/tendermint/pkg/consts"
 )
 
 // These constants were originally sourced from:
 // https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/consensus.md#constants
 const (
-	// NamespaceSize is the namespace size in bytes.
-	NamespaceSize = consts.NamespaceSize
+	// NamespaceSize is the size of a namespace (version + ID) in bytes.
+	NamespaceSize = 33
```
[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/appconsts/appconsts.go:12]: I can move consts from pkg/namespace here if that's what you're asking. I don't want this package to depend on pkg/namespace because that could lead to cyclic imports.

File: pkg/namespace/consts.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	// MaxReservedNamespace is lexicographically the largest namespace that is
+	// reserved for protocol use.
+	MaxReservedNamespace = MustNewV0([]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 255})
+
+	// TailPaddingNamespaceID is the namespace reserved for tail padding. All data
+	// with this namespace will be ignored.
+	TailPaddingNamespaceID = MustNewV0([]byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE})
+
+	// ParitySharesNamespaceID is the namespace reserved for erasure coded data.
+	ParitySharesNamespaceID = MustNewV0([]byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF})
```
Comment (rootulp) [inline pkg/namespace/consts.go:62]: [optional] we may consider removing the `ID` suffix from these constants because they are of type `Namespace`

Comment (rootulp) [reply pkg/namespace/consts.go]: done in https://github.com/celestiaorg/celestia-app/pull/1529/commits/a5a6ade324326e0153a80ac6bbd5848a7d03e221

File: proto/celestia/blob/v1/event.proto [hunk 4]
```diff
@@ -10,5 +10,8 @@ option go_package = "github.com/celestiaorg/celestia-app/x/blob/types";
 message EventPayForBlobs {
   string signer = 1;
   repeated uint32 blob_sizes = 2;
-  repeated bytes namespace_ids = 3;
+  // namespaces is a list of namespaces that the blobs in blob_sizes belong to.
+  // A namespace has length of 33 bytes where the first byte is the
+  // namespaceVersion and the subsequent 32 bytes are the namespaceID.
+  repeated bytes namespaces = 3;
```
Comment (evan-forbes) [inline proto/celestia/blob/v1/event.proto:16]: is this change required?

File: x/blob/types/payforblob.go [hunk 5]
```diff
@@ -195,14 +240,21 @@ func CreateCommitments(blobs []*Blob) ([][]byte, error) {
...
 	for _, blob := range blobs {
-		err := ValidateBlobNamespaceID(blob.NamespaceId)
+		if blob.NamespaceVersion > math.MaxUint8 {
+			return fmt.Errorf("namespace version %d is too large", blob.NamespaceVersion)
+		}
+		ns, err := appns.New(uint8(blob.NamespaceVersion), blob.NamespaceId)
+		if err != nil {
+			return err
+		}
+		err = ns.ValidateBlobNamespace()
```
Comment (evan-forbes) [inline x/blob/types/payforblob.go:257]: [optional comment/question]
could we combine the validation the the New method? If so, do we gain anything from that?

**Review by evan-forbes — COMMENTED** [review]
It's difficult to see the exact consensus breaking logic due to the number of changes in this PR, mind pointing them out to make sure I didn't miss anything? I had one blocking question, but I think this makes sense. I will review again after having some time to think over the important pieces here, nice work!

File: pkg/shares/testdata/sample-block.json [hunk 6]
```diff
@@ -3,38 +3,38 @@
 		"version": {
 			"block": 11
 		},
-		"chain_id": "fjOJHN",
+		"chain_id": "chain-id",
```
[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/shares/testdata/sample-block.json:6]: Context: https://github.com/celestiaorg/celestia-app/pull/1298 and https://github.com/celestiaorg/celestia-app/issues/1322

I think if we have test data that generates PFB txs then we can test `merge` using such data and can drop the sample-block.json

File: pkg/namespace/namespace.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+func New(version uint8, id []byte) (Namespace, error) {
+	err := validateVersion(version)
+	if err != nil {
+		return Namespace{}, err
+	}
+
+	err = validateID(version, id)
+	if err != nil {
+		return Namespace{}, err
+	}
```
Comment (evan-forbes) [inline pkg/namespace/namespace.go:23]: should this validation logic go in the validation method?

File: proto/celestia/blob/v1/event.proto (hunk 4 above)
Comment (rootulp) [reply proto/celestia/blob/v1/event.proto:16]: it's a name change so not strictly required but I think it's confusing to have a field called `namespace_ids` that contains the namespace version and namespace ID

File: pkg/namespace/namespace.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	if n.IsParityShares() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use parity shares namespace ID", n.Bytes())
+	}
+
+	if n.IsTailPadding() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use tail padding namespace ID", n.Bytes())
+	}
+
+	return nil
+}
```
Comment (evan-forbes) [inline pkg/namespace/namespace.go:85]: shouldn't we also be checking that only a portion of the namespace is used? is that in a different PR?

File: x/blob/types/blob_tx.go [hunk 9]
```diff
@@ -71,10 +72,19 @@ func ValidateBlobTx(txcfg client.TxEncodingConfig, bTx tmproto.BlobTx) error {
...
 
-	for i := range pfb.NamespaceIds {
-		// check that the metadata matches
-		if !bytes.Equal(bTx.Blobs[i].NamespaceId, pfb.NamespaceIds[i]) {
-			return ErrNamespaceMismatch.Wrapf("%v %v", bTx.Blobs[i].NamespaceId, pfb.NamespaceIds[i])
+	for i, ns := range pfb.Namespaces {
+		pfbNamespace, err := appns.From(ns)
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (rootulp) [reply x/blob/types/blob_tx.go:79]: 1. https://github.com/rootulp/celestia-app/blob/a5a6ade324326e0153a80ac6bbd5848a7d03e221/pkg/namespace/namespace.go#L55-L56
2. pfbNamespace is probably a confusing variable name. Is `namespace` clearer?

File: x/blob/types/payforblob.go (hunk 5 above)
Comment (rootulp) [reply x/blob/types/payforblob.go:257]: we could if we created a new function:
```go
NewBlobNamespace
```

I don't think it makes sense to enforce this validation on all usage of `New` because all reserved namespaces will fail the `ValidateBlobNamespace` check

File: pkg/namespace/namespace.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+// validateID returns an error if the provided id does not meet the requirements
+// for the provided version.
+func validateID(version uint8, id []byte) error {
+	if len(id) != NamespaceIDSize {
+		return fmt.Errorf("unsupported namespace id length: id %v must be %v bytes but it was %v bytes", id, NamespaceIDSize, len(id))
+	}
+
+	if version == NamespaceVersionZero && !bytes.HasPrefix(id, NamespaceVersionZeroPrefix) {
+		return fmt.Errorf("unsupported namespace id with version %v. ID %v must start with %v leading zeros", version, id, len(NamespaceVersionZeroPrefix))
+	}
```
Comment (evan-forbes) [inline pkg/namespace/namespace.go:104]: ahh I see now

File: pkg/namespace/namespace.go (hunk 7 above)
Comment (rootulp) [reply pkg/namespace/namespace.go:23]: It can but why should the `Namespace` type support creating a new namespace that violates the rules of version / ID? Then consumers need to call `validateBasic()` prior to using a Namespace. If we enforce these variants at construction time (i.e. `New()`) then users don't need to validate the namespace at each usage.

**Review by mojtaba-esk — APPROVED** [review]

File: proto/celestia/blob/v1/event.proto (hunk 4 above)
Comment (evan-forbes) [reply proto/celestia/blob/v1/event.proto:16]: I see! yeah, the change makes sense

File: x/blob/types/payforblob.go (hunk 5 above)
Comment (evan-forbes) [reply x/blob/types/payforblob.go:257]: I see thanks for the clarification

File: app/process_proposal.go [hunk 11]
```diff
@@ -45,8 +45,15 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponsePr
...
-			logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("invalid blob namespace %v", blob.NamespaceID))
+		namespace, err := appns.New(blob.NamespaceVersion, blob.NamespaceID)
+		if err != nil {
+			return abci.ResponseProcessProposal{
+				Result: abci.ResponseProcessProposal_REJECT,
+			}
+		}
+		err = namespace.ValidateBlobNamespace()
+		if err != nil {
+			logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("invalid blob namespace %v", namespace.Bytes()))
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply app/process_proposal.go:56]: ahh I see! I just missed that :sweat_smile:

File: x/blob/types/blob_tx.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply x/blob/types/blob_tx.go]: Attempted to clarify variable name in https://github.com/celestiaorg/celestia-app/pull/1529/commits/41c541386b43d993bbd32e92f37cf60d2c94f20e

File: pkg/namespace/namespace.go (hunk 8 above)
Comment (evan-forbes) [reply pkg/namespace/namespace.go:85]: thanks, I think I was just confused by why we had a separation of validation logic, but

> I don't think it makes sense to enforce this validation on all usage of New because all reserved namespaces will fail the ValidateBlobNamespace check

explains it

File: pkg/namespace/namespace.go (hunk 7 above)
Comment (evan-forbes) [reply pkg/namespace/namespace.go:23]: https://github.com/celestiaorg/celestia-app/pull/1529/files#r1150771472

File: x/blob/types/payforblob.go (hunk 5 above)
Comment (evan-forbes) [reply x/blob/types/payforblob.go:257]: out of curiosity, where are we using the new function to create reserved namespaces?

Comment (rootulp) [reply x/blob/types/payforblob.go:257]: 1. https://github.com/rootulp/celestia-app/blob/41c541386b43d993bbd32e92f37cf60d2c94f20e/app/process_proposal.go#L48
2. https://github.com/rootulp/celestia-app/blob/41c541386b43d993bbd32e92f37cf60d2c94f20e/pkg/shares/split_sparse_shares.go#L32
3. https://github.com/rootulp/celestia-app/blob/41c541386b43d993bbd32e92f37cf60d2c94f20e/x/blob/client/cli/payforblob.go#L39
+ usage in 3 more files

File: app/process_proposal.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply app/process_proposal.go:56]: I'm still working on a ProcessProposal test for this

File: pkg/namespace/consts.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	// NamespaceVersionZero is the first namespace version.
+	NamespaceVersionZero = uint8(0)
+
+	// NamespaceZeroPrefixSize is the number of `0` bytes that are prefixed to
+	// namespace IDs for version 0.
+	NamespaceVersionZeroPrefixSize = 22
+
+	// NamespaceVersionZeroIDSize is the number of bytes available for
+	// user-specified namespace ID in a namespace ID for version 0.
+	NamespaceVersionZeroIDSize = NamespaceIDSize - NamespaceVersionZeroPrefixSize
```
Comment (staheri14) [inline pkg/namespace/consts.go:28]: [suggestion-optional] what do you think about defining a type for Namespace version `0` e.g., an array of 10 bytes? it may make type conversions and boundary checks easier.

File: pkg/namespace/namespace.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+func MustNew(version uint8, id []byte) Namespace {
+	ns, err := New(version, id)
+	if err != nil {
+		panic(err)
+	}
+	return ns
+}
+
+// MustNewV0 returns a new namespace with version 0 and the provided id. This
+// function panics if the provided id is not exactly 10 bytes.
```
[the remark this answers is not in the document]
Comment (staheri14) [reply pkg/namespace/namespace.go:42]:
```suggestion
// function panics if the provided id is not exactly NamespaceVersionZeroIDSize bytes.
```

**Comment by staheri14** [comment]
Great work! added some quick suggestions above, will continue my review.

File: app/estimate_square_size_test.go [hunk 14]
```diff
@@ -159,7 +159,7 @@ func Test_estimateTxSharesUsed(t *testing.T) {
 		{"one tx", generateNormalTxs(1), 1},             // 1 tx is approximately 316 bytes which fits in 1 share
 		{"two txs", generateNormalTxs(2), 2},            // 2 txs is approximately 632 bytes which fits in 2 shares
 		{"ten txs", generateNormalTxs(10), 7},           // 10 txs is approximately 3160 bytes which fits in 7 shares
-		{"one hundred txs", generateNormalTxs(100), 67}, // 100 txs is approximately 31600 bytes which fits in 67 share
+		{"one hundred txs", generateNormalTxs(100), 68}, // 100 txs is approximately 31600 bytes which fits in 68 share
```
[the remark this answers is not in the document]
Comment (rootulp) [reply app/estimate_square_size_test.go:162]: good catch, will update.

File: pkg/namespace/consts.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	NamespaceVersionSize = appconsts.NamespaceVersionSize
+
+	// NamespaceIDSize is the size of a namespace ID in bytes.
+	NamespaceIDSize = appconsts.NamespaceIDSize
+
+	// NamespaceSize is the size of a namespace (version + ID) in bytes.
+	NamespaceSize = appconsts.NamespaceSize
+
+	// NamespaceVersionZero is the first namespace version.
+	NamespaceVersionZero = uint8(0)
```
[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/namespace/consts.go:20]: I'm not opposed. What is the benefit of using a type alias instead of using `uint8` directly?

Ref: https://yourbasic.org/golang/type-alias/ doesn't point out any strong benefits.

File: pkg/namespace/consts.go (hunk 12 above)
Comment (rootulp) [reply pkg/namespace/consts.go:28]: I'm not opposed to using a fixed length array instead of a byte slice but I don't expect it to make type conversions easier. Can you elaborate on why a fixed length array is a better choice than a slice here?

File: app/estimate_square_size_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply app/estimate_square_size_test.go]: resolved by https://github.com/celestiaorg/celestia-app/pull/1529/commits/c9fbfbbe3a290c8a96613d7431943827379f4160

File: app/process_proposal.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply app/process_proposal.go:56]: Spun out https://github.com/celestiaorg/celestia-app/issues/1558 b/c we want to merge this to unblock testing an unrelated PR

File: pkg/appconsts/appconsts.go [hunk 16]
```diff
@@ -1,16 +1,21 @@
...
-	// NamespaceSize is the namespace size in bytes.
-	NamespaceSize = consts.NamespaceSize
+	// NamespaveVersionSize is the size of a namespace version in bytes.
+	NamespaceVersionSize = 1
+
+	// NamespaceIDSize is the size of a namespace ID in bytes.
+	NamespaceIDSize = 32
+
+	// NamespaceSize is the size of a namespace (version + ID) in bytes.
+	NamespaceSize = NamespaceVersionSize + NamespaceIDSize
```
[the remark this answers is not in the document]
Comment (staheri14) [reply pkg/appconsts/appconsts.go:18]: I see, The way that it has been used in the `nmt` package is different, the `namespace ID` and `namespace` words are used interchangeably. For example, https://github.com/celestiaorg/nmt/blob/eded06696df88e15fcd8625047653957f548e383/nmt.go#L15. 
We don't have to use my suggested name i.e., `NamespaceIDVersionSize`,any name that resolves the potential confusion would work.

File: pkg/namespace/consts.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (staheri14) [reply pkg/namespace/consts.go:20]: Good question! For instance, if we were to allocate 16 bits for the version in the future, we could easily do so by updating the alias type. This change would immediately clarify all other places where the alias is used, making refactoring a simple and straightforward process.

[the remark this answers is not in the document]
Comment (staheri14) [reply pkg/namespace/consts.go:20]: At present, the reason for defining `NamespaceVersionZero` (or any other variable holding a version type as its value) as uint8 is unclear. It is unclear whether this is intended for mathematical operations with another uint8 value or if the version is supposed to be 8 bits long by default. This question may also arise in other parts of the code.

File: pkg/namespace/consts.go (hunk 12 above)
Comment (staheri14) [reply pkg/namespace/consts.go:28]: Sure, the main advantage is that we can avoid having to check the input size in every function call to ensure that it matches the expected length. By defining and initializing the variable as an array, we can safely pass it around the code without worrying about its size.

Comment (rootulp) [reply pkg/namespace/consts.go:28]: The namespace struct enforces an invariant for the length of a namespace so code that passes around namespaces don't need to check for length.

File: pkg/appconsts/appconsts.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/appconsts/appconsts.go:18]: I think the crux is that a namespace in celestia-app is being used as a namespace ID in the NMT repo so one option is to rename instances of namespace ID to namespace in the nmt repo.

File: pkg/namespace/consts.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (rootulp) [reply pkg/namespace/consts.go:20]: The three articles I read on it just now make it seem like the primary benefit is refactoring. So I agree with your point on refactoring but I think it comes at the cost of readability (subjective argument) because IMO this

```go
func NewCompactShareSplitter(ns appns.Namespace, shareVersion uint8) *CompactShareSplitter {
```

is slightly easier to read than this:

```go
func NewCompactShareSplitter(ns appns.Namespace, shareVersion appconsts.ShareVersion) *CompactShareSplitter {
```

because the latter forces you to hover or jump to another file to learn what underlying type a `ShareVersion` is. I'm a little hesitant to introduce them because:

> Type aliases are not meant for everyday use.

Ref: https://yourbasic.org/golang/type-alias/

> Type alias declaration wasn’t designed with everyday use in mind. 

Ref: https://gosolve.io/golang-type-alias/

but will open a follow-up issue if you have strong opinions about adopting it.
